File: src/main/python/AoC2017_04.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 4
#

import sys
from collections import Counter
from typing import Callable

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Input = list[str]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return list(input_data)

    def solve(
        self, inputs: list[str], strategy: Callable[[list[str]], bool]
    ) -> int:
        return sum(strategy(line.split()) for line in inputs)

    def part_1(self, inputs: list[str]) -> int:
        def has_no_duplicate_words(words: list[str]) -> bool:
            return Counter(words).most_common(1)[0][1] == 1

        return self.solve(inputs, has_no_duplicate_words)

    def part_2(self, inputs: list[str]) -> int:
        def has_no_anagrams(words: list[str]) -> bool:
            return len(words) == len(
                {tuple(sorted(Counter(word).items())) for word in words}
            )

        return self.solve(inputs, has_no_anagrams)

    @aoc_samples(
        (
            ("part_1", "aa bb cc dd ee", 1),
            ("part_1", "aa bb cc dd aa", 0),
            ("part_1", "aa bb cc dd aaa", 1),
            ("part_2", "abcde fghij", 1),
            ("part_2", "abcde xyz ecdab", 0),
            ("part_2", "a ab abc abd abf abj", 1),
            ("part_2", "iiii oiii ooii oooi oooo", 1),
            ("part_2", "oiii ioii iioi iiio", 0),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2017, 4)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/julia/AoC2017_04.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2017_04

include("aoc.jl")

function merge(d, key, value, func)
    d[key] = func(get(d, key, 0), value)
end

function hasNoDuplicateWords(line)
    d::Dict{String,Integer} = Dict()
    foreach(word -> merge(d, word, 1, +), split(line))
    return all(==(1), values(d))
end

function hasNoAnagrams(line)
    words = split(line)
    letterCounts::Set{Dict{Char,Int}} = Set()
    for word ∈ words
        d::Dict{Char,Integer} = Dict()
        foreach(ch -> merge(d, ch, 1, +), word)
        push!(letterCounts, d)
    end
    return length(words) == length(letterCounts)

end

function solve(input, strategy::Function)
    return count(line->strategy(line), input)
end

function part1(input)
    return solve(input, hasNoDuplicateWords)
end

function part2(input)
    return solve(input, hasNoAnagrams)
end

function samples()
    @assert part1(("aa bb cc dd ee",)) == 1
    @assert part1(("aa bb cc dd aa",)) == 0
    @assert part1(("aa bb cc dd aaa",)) == 1
    @assert part2(("abcde fghij",)) == 1
    @assert part2(("abcde xyz ecdab",)) == 0
    @assert part2(("a ab abc abd abf abj",)) == 1
    @assert part2(("iiii oiii ooii oooi oooo",)) == 1
    @assert part2(("oiii ioii iioi iiio",)) == 0
end

end # module AoC2017_04

aoc_main(@__FILE__, ARGS, 2017, 4)
